mips/ebase.S

```diff
@@ -29,59 +26,59 @@ _ebase:
         .local  invalid
         .local  pt_access
 
-tlb_refill:
-        # We don't want to generate TLB misses during tlb_refill.
+# TLB miss handler must avoid generating TLB miss during PT search.
+# 
+# Notes for the reader:
+#  - tlbr instruction overwrites EntryHi, so we have restore it,
+#  - handler does not fit into 256B, so it is splitted into two parts.
 
-        # Special case: an access was made to the PT from outside tlb_refill.
-        # This needs to be checked and handled separately.
+tlb_refill:
+        # Consider special case when an access to page table region was from
+        # the kernel (namely pmap module). As UPD & KPD are not mapped within
+        # page table we have to deal with it separately :(
 
         mfc0    k0, C0_BADVADDR
-        srl     k1, k0, 22
-        addi    k1, k1, -(PT_BASE >> 22)
+        srl     k1, k0, PDE_SHIFT
+        addi    k1, k1, -(PT_BASE >> PDE_SHIFT)
         beqz    k1, pt_access
         nop
 
         mfc0    k0, C0_CONTEXT
-        sra     k0, 1                                   # k0 = address of PTE of even page from the pair
+        sra     k0, 1
 
-        # Check if it's safe to access the PT (is there a valid entry in the TLB?)
-        mfc0    k1, C0_ENTRYHI                          # Save original EntryHi for later.
-        ins     k0, k1, 0, C0_ENTRYHI_VPN2_SHIFT        # Copy current ASID.
+        # Is it safe to access the PT? (is there a valid entry in the TLB?)
+        mfc0    k1, C0_ENTRYHI  # Save original EntryHi
+        ins     k0, k1, 0, C0_ENTRYHI_VPN2_SHIFT # Copy current ASID
         mtc0    k0, C0_ENTRYHI
         ehb
         tlbp
         ehb
-        mtc0    k1, C0_ENTRYHI                          # Restore original EntryHi.
+        mtc0    k1, C0_ENTRYHI  # Restore original EntryHi
         mfc0    k0, C0_INDEX
-        # Index < 0 means there was no entry with matching VPN2.
-        bltz    k0, miss
+        bltz    k0, miss        # Index < 0 => no entry with matching VPN2
         nop
+
 maybe_hit:
-        # We found an entry with a matching VPN2, but that doesn't mean that it's valid!
-        # We need to check it.
+        # We found an entry with a matching VPN2, but that doesn't mean that
+        # it's valid! We need to check it.
 
-        # Read the entry with matching VPN2 (note: tlbr overwrites EntryHi!)
-        tlbr
-        mtc0    k1, C0_ENTRYHI                  # Restore original EntryHi.
+        tlbr                    # Read the entry with matching VPN2
+        ehb
```

Comment:
Programming the MIPS32® 24KETM Core Family, 5.1 Hazard barrier instructions:
> Since some implementations work by holding up execution of all instructions after the barrier, it’s preferable to place the barrier just before the consumer, not just after the producer.

In this case the producer is `tlbr`, the consumers are `mfc0` from `C0_ENTRYLO0/1`. There is no hazard between `tlbr` and `mtc0` or `mfc0` from `C0_CONTEXT`.

Reply:
Thanks for digging it up! This note is so imporant I'll put in into a comment.